### PR TITLE
use include_bytes! for target and definitions

### LINKED ISSUE #22: Pathological compile times
When you try to include large (> 10MB) files, this tends to blow `rustc`'s compile times and memory usage massively out of proportion. With a 10MB file full of random binary data it ended up taking about half an hour and 20GB of RAM to compile...

I'll need to check whether `include_bytes!()` and `include_str!()` also suffer from this, and maybe have a look at their implementations.

### PR DESCRIPTION
**Pull request by arBmind**
fixes #22 
* 1.5kb takes ~4 seconds (11s for release)
* 80mb takes ~9 seconds (18s for release)

issues:
* removing a file will not trigger a `build.rs` run and fails to compile `include_bytes!` fails to find the file.
* adding files will not trigger a `build.rs` run and fails to include the new file.

### CODE REVIEW

File: src/serializer.rs [hunk 1]
```diff
@@ -53,325 +45,48 @@ impl<W> Serializer<W>
...
-                            self.extend_contents(d);
-                            Some(DirEntry::Dir(d))
-                        } else {
-                            entry
-                        }
-                    }
-                }")?;
-
+    fn write_std_definitions(&mut self) -> Result<&mut Self> {
+        self.writer.write(include_bytes!("serialized_std_definitions.rs"))?;
```
Comment (Michael-F-Bryan) [inline src/serializer.rs]: Good idea removing the string literals and replacing them with `include_bytes!()` :+1:

File: src/files.rs [hunk 2]
```diff
@@ -44,38 +38,33 @@ impl Locatable for File {
...
-        file.contents().unwrap().read_to_end(&mut got).unwrap();
-        assert_eq!(got, file_contents);
-        assert_eq!(file.name(), path);
+        assert_eq!(file.name(), f.path());
+        f.close().unwrap();
     }
 
     #[test]
     fn file_only_works_on_files() {
-        let t = TempDir::new("blah").unwrap();
```
Comment (Michael-F-Bryan) [inline src/files.rs]: I know it's a massive edge case and I doubt anyone would ever have a file called `blah` in their working directory, but do think it's worth inserting an `assert!(!t.exists())`?

**Comment by Michael-F-Bryan** [comment]
Looking through the changes you added, am I correct in assuming that the long compile times and excessive RAM usage were because the file contents were originally being written into the source code as a byte array instead of with `include_bytes!()`?

File: src/files.rs (hunk 2 above)
Comment (arBmind) [reply src/files.rs]: I never understood what the code was trying to achieve with creating a TempDir. Now you mention that and I realize that I should not have removed this.

**Comment by arBmind** [comment]
Actually I never tried the `include_dir` with large files before this change.
I created this patch, because I wanted to try to use `include_bytes!` or `include_str!` macros instead of creating large rust files. It worked very well, so I finished the pull request. Only after I created the PR I found the issue.

1mb binary become 4-5mb of rust source. The compiler will have to parse this again, creating AST objects for each array entry.

File: src/serializer.rs (hunk 1 above)
Comment (arBmind) [reply src/serializer.rs]: The output formatting is now correct as well ☮️

File: src/files.rs (hunk 2 above)
Comment (Michael-F-Bryan) [reply src/files.rs]: I think I misread the test when writing that previous comment. The test creates a temporary directory and then makes sure that executing `include_file()` with a directory is an error. So it looks like you might want to put that `TempDir` back in. 

If you can think of a more descriptive name for the test feel free to change it. It may help future developers understand what it's doing.

File: src/files.rs [hunk 3]
```diff
@@ -62,8 +63,7 @@ mod tests {
 
     #[test]
     fn file_only_works_on_files() {
```
Comment (Michael-F-Bryan) [inline src/files.rs]: Would `cant_use_include_file_on_directories` be a better name?

Comment (arBmind) [reply src/files.rs]: I think this discussion is discussion is valid, but not an essential part of this PR.

Naming is part of the higher arts of testing. To tell the truth I never read the test name. I was on this file because the other test broke. Removing the content parts left me in the mood to remove more.

I would suggest to avoid negations in names and aim for the shortest name. `reject_directory` or `err_on_directory`. This also makes it obvious that a test for non existing files is missing. On the other hand this code now is so trivial, that I would argue that you have enough tests as is.

The more interesting discussion is if this file is still needed. Basically File is a PathBuf. PathBuf allows to check the file existence in one line. So we have a file with 65 lines of code that gain us basically nothing for the rest of this crate.

Comment (Michael-F-Bryan) [reply src/files.rs]: You raise a good point. I'll create a separate issue regarding the test though seeing as it isn't really relevant to the PR.

**Comment by Michael-F-Bryan** [comment]
I've created #25 to deal with the two issues you mention in the PR's description. I also opened #26 to re-evaluate whether `File` is still necessary, seeing as it's the equivalent of a `PathBuf`.

It looks like this is good to merge though.

**Comment by Michael-F-Bryan** [comment]
A new version is now up on crates.io.
